### ISSUE TITLE
TS2322 not work on some strange situations

### ISSUE THREAD

**Issue by FlareStone** [microsoft/TypeScript#50784]
# Bug Report

### 🔎 Search Terms
type or, adding undefined, TS2322

### 🕗 Version & Regression Information
Maybe it is from the beginning

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true&ts=4.9.0-dev.20220915#code/C4TwDgpgBAglC8UDeAoK6oDsCGBbCAXFAM7ABOAlpgOYoC+KokUAQgsmhjvkaZTZ3QUAJkUwBXXACMIZeikbhoAYXZwAPqwUBjAPaZSUbEVWJUGLHkJQA5AEYbAGkFQRRcZmEQAZlQjDnBhQ9A2AoKRN2cy4rInsnFzcoDy9fTH9nCwA3bAAbcWsAJkCUIA)

### 💻 Code

```ts
type A = {
    name: string
}
type B = {
    name: string
    id: number
}

type C = A | B

const a: C = {
    name: '1',
    id: undefined, // expect an error
}

const b: C = {
    name: '1',
    id: undefined, // an ts2322 error
    value: 2,
}

```

### 🙁 Actual behavior

The var a does not report any error, but the var b report error "Type 'undefined' is not assignable to type 'number'(2322)".
Before v3.7.5, it will report error "Type '{ name: string; id: undefined; value: number; }' is not assignable to type 'C'.
  Object literal may only specify known properties, and 'value' does not exist in type 'C'.(2322)" on b.
But I think the error should occur on var a too.

### 🙂 Expected behavior
The error reported on b should also be reported on var a.

**Comment by MartinJohns** [comment]
Excess property checks are not working for unions without a discriminant. See #20863.
